### PR TITLE
oci/build: Specify ArtifactType in the correct place

### PR DESCRIPTION
**Pull request by blanquicet**
# Specify ArtifactType in the correct place when building the image

As discussed in https://github.com/inspektor-gadget/inspektor-gadget/pull/2177, the ArtifactType must be specified when config.mediaType is set MediaTypeEmptyJSON. In our case, when the metadata file is not provided. However, such PR added the ArtifactType inside the config and not to the general manifest as described in [1].

Fixes: https://github.com/inspektor-gadget/inspektor-gadget/commit/1b04865d86c95ffaa9631d1bd73b43c6a2c43337

[1] https://github.com/opencontainers/image-spec/blob/f5f87016de46439ccf91b5381cf76faaae2bc28f/manifest.md?plain=1#L210

### Before this PR
```bash
$ IG_EXPERIMENTAL=true go run --exec "sudo -E" ../../cmd/ig/... image build -t ghcr.io/blanquicet/gadget/trace_exec:latest .
$ IG_EXPERIMENTAL=true go run --exec "sudo -E" ../../cmd/ig/... image push ghcr.io/blanquicet/gadget/trace_exec:latest
$ crane manifest ghcr.io/blanquicet/gadget/trace_exec:latest | jq .
... Get SHA of one of the manifests ...
$ crane manifest ghcr.io/blanquicet/gadget/trace_exec:latest@sha256:ec03ea97b5b0368a823a472f54fcbbf07312501b8b298a54ca9f1565f73b50c1 | jq .
{
  "schemaVersion": 2,
  "config": {
    "mediaType": "application/vnd.oci.empty.v1+json",
    "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
    "size": 2,
    "data": "e30=",
    "artifactType": "application/vnd.gadget.ebpf.program.v1+binary"
  },
  "layers": [
    {
      "mediaType": "application/vnd.gadget.ebpf.program.v1+binary",
      "digest": "sha256:2f16368166aee02a9a44768abe85dec3b0340392927e4a1f55269df09ea0c87b",
      "size": 86904
    }
  ]
}
```

### After this PR

The new result now looks like the example provided by [image-spec](https://github.com/opencontainers/image-spec/blob/f5f87016de46439ccf91b5381cf76faaae2bc28f/manifest.md?plain=1#L210):
```bash
$ crane manifest ghcr.io/blanquicet/gadget/trace_exec:latest@sha256:ec03ea97b5b0368a823a472f54fcbbf07312501b8b298a54ca9f1565f73b50c1 | jq .
{
  "schemaVersion": 2,
  "artifactType": "application/vnd.gadget.ebpf.program.v1+binary",
  "config": {
    "mediaType": "application/vnd.oci.empty.v1+json",
    "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
    "size": 2,
    "data": "e30="
  },
  "layers": [
    {
      "mediaType": "application/vnd.gadget.ebpf.program.v1+binary",
      "digest": "sha256:e0725896ea02f1ff09e58b85d55de0453dcfcef94b7c51e907c54faa35df60cf",
      "size": 149960
    }
  ]
}
```

### CODE REVIEW

**Review by mqasimsarfraz — APPROVED** [review]
Great Catch! Out of curiosity how did you catch this? Perhaps we can consider to [create tests](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/integration/components/image) for our manifest validation in future.

Tested with steps mentioned in https://github.com/inspektor-gadget/inspektor-gadget/pull/2177 and no problems in pushing the image as well.

**Comment by mauriciovasquezbernal** [comment]
Does it solve https://github.com/inspektor-gadget/inspektor-gadget/issues/2289?

**Comment by blanquicet** [comment]
> Great Catch! Out of curiosity how did you catch this?
>
While reviewing https://github.com/inspektor-gadget/inspektor-gadget/pull/2512#discussion_r1510583296, I noticed it was looking at the `ArtifactType` and I noticed there was a "general" `ArtifactType` for the whole manifest and another for the config.

> Perhaps we can consider to [create tests](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/integration/components/image) for our manifest validation in future.
> 
Agree! https://github.com/inspektor-gadget/inspektor-gadget/issues/2577

> Tested with steps mentioned in #2177 and no problems in pushing the image as well.
>
Thanks!

**Comment by blanquicet** [comment]
> Does it solve #2289?

Yep. It does! I test it with trace_open in my ACR registry:

```bash
jose trace_open git:(main *)$ ls
program.bpf.c
```

### Before this PR

```bash
jose trace_open git:(main *)$ IG_EXPERIMENTAL=true go run --exec "sudo -E" ../../cmd/ig/... image build -t joseregistry.azurecr.io/trace_open:latest .
INFO[0000] Experimental features enabled
Successfully built joseregistry.azurecr.io/trace_open:latest@sha256:4dd9d2a330a85d15eb69b75a261dda17045d5c3fb68287e16030ba429af779b6

jose trace_open git:(main *)$ IG_EXPERIMENTAL=true go run --exec "sudo -E" ../../cmd/ig/... image push joseregistry.azurecr.io/trace_open:latest
INFO[0000] Experimental features enabled
Pushing joseregistry.azurecr.io/trace_open:latest...
Error: pushing gadget: copying to remote repository: PUT "https://joseregistry.azurecr.io/v2/trace_open/manifests/sha256:5b0c1ddffd0b687226fc850a1a3b3d514310f4ebbc7e0d148e9ce35c4beeb831": response status code 400: manifest invalid: manifest invalid: map[Reason:using empty config descriptor requires artifactType to be defined]
exit status 1
```

### After this PR

```bash
jose trace_open git:(main *)$ IG_EXPERIMENTAL=true go run --exec "sudo -E" ../../cmd/ig/... image build -t joseregistry.azurecr.io/trace_open:latest .
INFO[0000] Experimental features enabled
Successfully built 
joseregistry.azurecr.io/trace_open:latest@sha256:b78f41657ea1e0492122fba1898728f19a2610cb7cff93b67009610ba41407ff

jose trace_open git:(jose/empty-descriptor *)$ IG_EXPERIMENTAL=true go run --exec "sudo -E" ../../cmd/ig/... image push joseregistry.azurecr.io/trace_open:latest
INFO[0000] Experimental features enabled
Pushing joseregistry.azurecr.io/trace_open:latest...
Successfully pushed joseregistry.azurecr.io/trace_open:latest@sha256:b78f41657ea1e0492122fba1898728f19a2610cb7cff93b67009610ba41407ff
```

File: pkg/oci/build.go [hunk 1]
```diff
@@ -193,6 +192,10 @@ func createManifestForTarget(ctx context.Context, target oras.Target, metadataFi
 		return ocispec.Descriptor{}, fmt.Errorf("creating and pushing eBPF descriptor: %w", err)
 	}
 
+	// artifactType must be only set when the config.mediaType is set to
+	// MediaTypeEmptyJSON. In our case, when the metadata file is not provided:
+	// https://github.com/opencontainers/image-spec/blob/f5f87016de46439ccf91b5381cf76faaae2bc28f/manifest.md?plain=1#L170
+	var artifactType string
```
Comment (mauriciovasquezbernal) [inline pkg/oci/build.go:198]: I'm 6 months late: Is it really true? The third example in https://github.com/opencontainers/image-spec/blob/main/manifest.md#guidelines-for-artifact-usage has both defined, config.mediaType and artifactType. Can't we set artifactType in all cases?

Comment (mauriciovasquezbernal) [reply pkg/oci/build.go:198]: > artifactType must be **only** set when the config.mediaType is set to MediaTypeEmptyJSON

I think the "only" in the sentence above implies that artifactType can only be set when config.mediaType is empty.

> BTW, why did you come back to this discussion? Did you find any issue?

I found while reading more to implement #3372 and I was curious about it as my understanding was that we can set it on all cases (and we agree on this one)

Comment (blanquicet) [reply pkg/oci/build.go:198]: > > artifactType must be **only** set when the config.mediaType is set to MediaTypeEmptyJSON
> 
> I think the "only" in the sentence above implies that artifactType can only be set when config.mediaType is empty.
> 
Nope. The "only" in the sentence above implies that artifactType ~can~ must only be set when config.mediaType is empty, which gives you the possibility to do it whenever you want, but you must do it only in that case... Anyway, we understood each other 😅

> > BTW, why did you come back to this discussion? Did you find any issue?
> 
> I found while reading more to implement #3372 and I was curious about it as my understanding was that we can set it on all cases (and we agree on this one)
>
OK 🙂

Comment (mauriciovasquezbernal) [reply pkg/oci/build.go:198]: Opened #3469

**Pull request merged**